src/storage/kv/transaction.rs

```diff
@@ -0,0 +1,1336 @@
+        // If the transaction is read-only, return an error.
+        if self.mode.is_read_only() {
+            return Err(Error::TransactionReadOnly);
+        }
+
+        // If there are no pending writes, there's nothing to commit, so return early.
+        if self.write_set.is_empty() {
+            return Ok(());
+        }
+
+        // TODO: Use a commit pipeline to avoid blocking calls.
+        // Lock the oracle to serialize commits to the transaction log.
+        let oracle = self.store.oracle.clone();
+        let _lock = oracle.write_lock.lock();
+
+        // Prepare for the commit by getting a transaction ID and a commit timestamp.
+        let (tx_id, commit_ts) = self.prepare_commit()?;
+
+        // Add transaction records to the transaction log.
+        self.add_to_transaction_log(tx_id, commit_ts)?;
+
+        // Commit the changes to the store index.
+        self.commit_to_index(tx_id, commit_ts)?;
+
+        // Update the oracle to indicate that the transaction has been committed up to the given transaction ID.
+        oracle.committed_upto(tx_id);
+
+        // Mark the transaction as closed.
+        self.closed = true;
+
+        Ok(())
+    }
+
+    /// Prepares for the commit by assigning commit timestamps and preparing records.
+    fn prepare_commit(&mut self) -> Result<(u64, u64)> {
+        let oracle = self.store.oracle.clone();
+        let tx_id = oracle.new_commit_ts(self)?;
+        let commit_ts = self.assign_commit_ts();
+        Ok((tx_id, commit_ts))
+    }
+
+    /// Assigns commit timestamps to transaction entries.
+    fn assign_commit_ts(&mut self) -> u64 {
+        let commit_ts = now();
+        self.write_set.iter_mut().for_each(|(_, entry)| {
+            entry.ts = commit_ts;
+        });
+        commit_ts
+    }
+
+    /// Adds transaction records to the transaction log.
+    fn add_to_transaction_log(&mut self, tx_id: u64, commit_ts: u64) -> Result<u64> {
+        let current_offset = self.store.clog.read().offset()?;
+        let entries: Vec<Entry> = self.write_set.values().cloned().collect();
+        let tx_record = TxRecord::new_with_entries(entries, tx_id, commit_ts);
+        tx_record.encode(
+            &mut self.buf,
+            current_offset,
+            &mut self.committed_values_offsets,
+        )?;
+
+        let mut clog = self.store.clog.write();
+        let (tx_offset, _) = clog.append(self.buf.as_ref())?;
+        Ok(tx_offset)
+    }
+
+    /// Commits transaction changes to the store index.
+    fn commit_to_index(&mut self, tx_id: u64, commit_ts: u64) -> Result<()> {
+        let mut index = self.store.indexer.write();
+        let mut kv_pairs = self.build_kv_pairs(tx_id, commit_ts);
+
+        index.bulk_insert(&mut kv_pairs)?;
+        Ok(())
+    }
+
+    /// Builds key-value pairs from the write set.
+    fn build_kv_pairs(&self, tx_id: u64, commit_ts: u64) -> Vec<KV<VariableKey, Bytes>> {
+        let mut kv_pairs: Vec<KV<VariableKey, Bytes>> = Vec::new();
```

Comment:
Is the type hint necessary here? Unless I'm missing something, the compiler should be able to infer it.

```suggestion
        let mut kv_pairs = Vec::new();
```

Reply:
You are right, not necessary at all